File: searchlib/src/main/java/com/yahoo/searchlib/expression/MultiArgPredicateNode.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.expression;

import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * Protocol class for descendants of {@code FilterExpressionNode} that has several {@code FilterExpressionNode} as children.
 *
 * @author johsol
 */
public abstract class MultiArgPredicateNode extends FilterExpressionNode {

    public static final int classId = registerClass(0x4000 + 175, MultiArgPredicateNode.class);
    private List<FilterExpressionNode> args = new ArrayList<>();

    public MultiArgPredicateNode() {}

    public MultiArgPredicateNode(List<FilterExpressionNode> args) {
        this.args = args;
    }

    public final MultiArgPredicateNode addArg(FilterExpressionNode arg) {
        Objects.requireNonNull(arg, "Can not add null arg"); // throws NullPointerException
        args.add(arg);
        return this;
    }

    public FilterExpressionNode getArg(int i) {
        return args.get(i);
    }

    public int getNumArgs() {
        return args.size();
    }

    public Optional<List<FilterExpressionNode>> getArgs() {
        return Optional.ofNullable(args);
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        int numArgs = args.size();
        buf.putInt(null, numArgs);
        for (FilterExpressionNode node : args) {
            serializeOptional(buf, node);
        }
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        args.clear();
        int numArgs = buf.getInt(null);
        for (int i = 0; i < numArgs; i++) {
            FilterExpressionNode node = (FilterExpressionNode)deserializeOptional(buf);
            args.add(node);
        }
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("args", args);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MultiArgPredicateNode rhs = (MultiArgPredicateNode)o;
        return args.equals(rhs.args);
    }

    @Override
    public int hashCode() {
        int ret = 0;
        for (FilterExpressionNode node : args) {
            ret += node.hashCode();
        }
        return ret;
    }
}

File: searchlib/src/vespa/searchlib/expression/multi_arg_predicate_node.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "multi_arg_predicate_node.h"
#include "resultvector.h"

namespace search::expression {

using vespalib::Deserializer;
using vespalib::Serializer;

IMPLEMENT_IDENTIFIABLE_ABSTRACT_NS2(search, expression, MultiArgPredicateNode, FilterPredicateNode);

MultiArgPredicateNode::MultiArgPredicateNode() noexcept = default;

MultiArgPredicateNode::~MultiArgPredicateNode() = default;

MultiArgPredicateNode::MultiArgPredicateNode(const std::vector<FilterPredicateNode::IP>& input) {
    for (const auto& node : input) {
        _args.emplace_back(node->clone());
    }
}

void MultiArgPredicateNode::visitMembers(vespalib::ObjectVisitor& visitor) const {
    visit(visitor, "args", _args);
}

void MultiArgPredicateNode::selectMembers(const vespalib::ObjectPredicate& predicate,
                                       vespalib::ObjectOperation& operation) {
    FilterPredicateNode::selectMembers(predicate, operation);
    for(auto& _arg : _args) {
        _arg->select(predicate, operation);
    }
}

} // namespace search::expression
